File: docs/scripts/i18n-segments.py
import json,re,sys,os
S='/private/tmp/claude-501/-Users-bytedance-projects-lynx-stack-docs-all-in-one/940f018a-b91e-4da2-bdb4-5c14e4908668/scratchpad/'
FENCE=re.compile(r'(```[^\n]*\n[\s\S]*?\n```)')
def segments(en):
    parts=FENCE.split(en)
    return parts  # even idx = prose, odd idx = fence
def dump(pkg):
    d=json.load(open(f'api-data/zh/{pkg}.json')); lines=[]
    for k,v in d.items():
        if v.get('text'): continue
        for i,seg in enumerate(segments(v['en'])):
            if i%2==0 and seg.strip():
                lines.append(f"{k}\t{i}\t{seg.strip().replace(chr(10),' ⏎ ')}")
    open(S+f'seg-{pkg}.tsv','w').write('\n'.join(lines)); return len(lines), sum(len(l) for l in lines)
def apply(pkg, tsv):
    d=json.load(open(f'api-data/zh/{pkg}.json'))
    tr={}
    for line in open(tsv, encoding='utf8', newline='\n'):
        if not line.strip() or line.startswith('#'): continue
        k,i,t=line.rstrip('\n').split('\t',2); tr.setdefault(k,{})[int(i)]=t.replace(' ⏎ ','\n')
    n=0
    for k,segs in tr.items():
        if k not in d or d[k].get('text'): continue
        parts=segments(d[k]['en'])
        for i,t in segs.items():
            if i<len(parts): parts[i]=('\n\n' if i>0 else '')+t+('\n\n' if i<len(parts)-1 else '')
        d[k]['text']=''.join(parts).strip(); d[k].pop('en',None); n+=1
    json.dump(d,open(f'api-data/zh/{pkg}.json','w'),indent=2,ensure_ascii=False); open(f'api-data/zh/{pkg}.json','a').write('\n')
    return n
def apply_dir(locale_dir):
    import glob
    groups={}
    for f in sorted(glob.glob(os.path.join(locale_dir,'seg-*.zh.tsv'))):
        pkg=re.sub(r'\.part\d+$','',os.path.basename(f)[4:-7])
        groups.setdefault(pkg,[]).append(f)
    for pkg,files in groups.items():
        merged=os.path.join(locale_dir,f'.merged-{pkg}.tsv')
        with open(merged,'w',encoding='utf8') as out:
            for f in files: out.write(open(f,encoding='utf8',newline='\n').read().replace('\r','').rstrip('\n')+'\n')
        print(f'{pkg:34s} applied {apply(pkg,merged)}')
        os.remove(merged)

if __name__=='__main__':
    cmd=sys.argv[1]
    if cmd=='apply-dir':
        apply_dir(sys.argv[2])
    elif cmd=='dump':
        for p in sys.argv[2:]:
            n,c=dump(p); print(f"{p:34s} {n:4d} segments {c:6d} chars")
    else:
        pkg,tsv=sys.argv[2],sys.argv[3]; print(pkg,'applied',apply(pkg,tsv))

File: docs/scripts/i18n-segments.ts
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { mkdirSync, readFileSync, readdirSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { hashText } from './render.ts';
import type { Translations } from './render.ts';

const ZH = join(dirname(fileURLToPath(import.meta.url)), '../api-data/zh');
const FENCE = /(```[^\n]*\n[\s\S]*?\n```)/;
const NEWLINE = ' ⏎ ';

function load(pkg: string): Translations {
  return JSON.parse(
    readFileSync(join(ZH, `${pkg}.json`), 'utf8'),
  ) as Translations;
}

function save(pkg: string, translations: Translations): void {
  writeFileSync(
    join(ZH, `${pkg}.json`),
    `${JSON.stringify(translations, null, 2)}\n`,
  );
}

function dump(dir: string, pkgs: string[]): void {
  mkdirSync(dir, { recursive: true });
  for (const pkg of pkgs) {
    const lines: string[] = [];
    for (const [key, t] of Object.entries(load(pkg))) {
      if (t.en === undefined) continue;
      t.en.split(FENCE).forEach((segment, i) => {
        if (i % 2 === 0 && segment.trim()) {
          lines.push(
            `${key}\t${i}\t${segment.trim().replaceAll('\n', NEWLINE)}`,
          );
        }
      });
    }
    writeFileSync(join(dir, `seg-${pkg}.tsv`), lines.join('\n'));
    console.info(`${pkg}: ${lines.length} segments`);
  }
}

function apply(dir: string): void {
  const byPackage = new Map<string, Map<string, Map<number, string>>>();
  const files = readdirSync(dir).filter(f => /^seg-.+\.zh\.tsv$/.test(f));
  for (const file of files.sort()) {
    const pkg = file.slice('seg-'.length, -'.zh.tsv'.length).replace(
      /\.part\d+$/,
      '',
    );
    const segments = byPackage.get(pkg)
      ?? new Map<string, Map<number, string>>();
    byPackage.set(pkg, segments);
    const text = readFileSync(join(dir, file), 'utf8').replaceAll('\r', '');
    for (const line of text.split('\n')) {
      if (!line.trim() || line.startsWith('#')) continue;
      const a = line.indexOf('\t');
      const b = line.indexOf('\t', a + 1);
      const key = line.slice(0, a);
      const byIndex = segments.get(key) ?? new Map<number, string>();
      segments.set(key, byIndex);
      byIndex.set(
        Number(line.slice(a + 1, b)),
        line.slice(b + 1).replaceAll(NEWLINE, '\n'),
      );
    }
  }
  for (const [pkg, segments] of byPackage) {
    const translations = load(pkg);
    let applied = 0;
    for (const [key, byIndex] of segments) {
      const t = translations[key];
      if (t?.en === undefined) continue;
      const parts = t.en.split(FENCE);
      for (const [i, text] of byIndex) {
        if (i >= parts.length) continue;
        parts[i] = (i > 0 ? '\n\n' : '') + text
          + (i < parts.length - 1 ? '\n\n' : '');
      }
      translations[key] = {
        hash: hashText(t.en),
        text: parts.join('').trim(),
      };
      applied++;
    }
    save(pkg, translations);
    console.info(`${pkg}: ${applied} applied`);
  }
}

const [command, dir, ...pkgs] = process.argv.slice(2);
if (command === 'dump' && dir) {
  dump(dir, pkgs);
} else if (command === 'apply' && dir) {
  apply(dir);
} else {
  console.error(
    'Usage: node scripts/i18n-segments.ts dump <dir> <package>... | apply <dir>',
  );
  process.exitCode = 1;
}
